src/sinks/util/sink.rs

```diff
@@ -0,0 +1,1193 @@
+        acker: Acker,
+        exec: E,
+    ) -> Self {
+        let service = ServiceSink::new(service, acker);
+
+        Self {
+            service,
+            batch,
+            settings,
+            linger: None,
+            closing: false,
+            exec,
+            _pd: PhantomData,
+        }
+    }
+
+    fn should_send(&mut self) -> bool {
+        self.closing || self.batch.len() >= self.settings.size || self.linger_elapsed()
+    }
+
+    fn linger_elapsed(&mut self) -> bool {
+        if let Some(delay) = &mut self.linger {
+            delay.poll().expect("timer error").is_ready()
+        } else {
+            false
+        }
+    }
+}
+
+impl<S, B, Request, E> Sink for BatchSink<S, B, Request, E>
+where
+    S: Service<Request>,
+    S::Future: Send + 'static,
+    S::Error: Into<crate::Error> + Send + 'static,
+    S::Response: fmt::Debug,
+    B: Batch<Output = Request>,
+    E: Executor,
+{
+    type SinkItem = B::Input;
+    type SinkError = crate::Error;
+
+    fn start_send(&mut self, item: Self::SinkItem) -> StartSend<Self::SinkItem, Self::SinkError> {
+        if self.batch.len() >= self.settings.size {
+            trace!("batch full.");
+            self.poll_complete()?;
+
+            if self.batch.len() > self.settings.size {
+                debug!(message = "Buffer full; applying back pressure.", size = %self.settings.size, rate_limit_secs = 10);
+                return Ok(AsyncSink::NotReady(item));
+            }
+        }
+
+        if self.batch.len() == 0 {
+            trace!("Creating new batch.");
+            // We just inserted the first item of a new batch, so set our delay to the longest time
+            // we want to allow that item to linger in the batch before being flushed.
+            let deadline = Instant::now() + self.settings.timeout;
+            self.linger = Some(Delay::new(deadline));
+        }
+
+        self.batch.push(item);
+
+        Ok(AsyncSink::Ready)
+    }
+
+    fn poll_complete(&mut self) -> Poll<(), Self::SinkError> {
+        loop {
+            if self.batch.is_empty() {
+                trace!("no batches; driving service to completion.");
+                return self.service.poll_complete();
+            } else {
+                // We have data to send, so check if we should send it and either attempt the send
+                // or return that we're not ready to send. If we send and it works, loop to poll or
+                // close inner instead of prematurely returning Ready
+                if self.should_send() {
+                    try_ready!(self.service.poll_ready());
+
+                    trace!("Service ready; Sending batch.");
+                    let batch = self.batch.fresh_replace();
+
+                    let batch_size = batch.num_items();
+                    let request = batch.finish();
+
+                    let fut = self.service.call(request, batch_size);
+
+                    self.exec.spawn(fut).unwrap();
```

Comment:
Might be good to `expect` instead of `unwrap` so we can explain why it shouldn't fail.

Reply:
:+1: Yeah, this should never fail ever basically. Since, we always spawn through tokio and tests will not pass if this fails but I can add an expect.